### PR TITLE
support multi-dim reward for SAC with discrete actions

### PR DESCRIPTION
**Pull request by hnyu**
Support multi-dim reward and reward weights for discrete SAC. 

Related to PR https://github.com/HorizonRobotics/alf/pull/683 where only continuous actions were supported. 

TODO: multi-dim reward support for Mixed action type.

### CODE REVIEW

File: alf/algorithms/sac_algorithm.py [hunk 1]
```diff
@@ -623,35 +655,45 @@ def _select_q_value(self, action, q_values):
...
 
         if self._act_type == ActionType.Discrete:
             critics = self._select_q_value(rollout_info.action, critics)
-            target_critics = self._select_q_value(
-                action, target_critics.unsqueeze(dim=1))
+            # [B, num_actions] -> [B, num_actions, reward_dim]
+            probs = common.expand_dims_as(action_distribution.probs,
+                                          target_critics)
+            # [B, reward_dim]
+            target_critics = torch.sum(probs * target_critics, dim=1)
```
Comment (hnyu) [inline alf/algorithms/sac_algorithm.py:696]: Changed from sampling estimate to weighted average. Now the target critics should be more accurate.

File: setup.py [hunk 2]
```diff
@@ -44,7 +44,7 @@
         'sphinx-autodoc-typehints',
         'sphinxcontrib-napoleon==0.7',
         'sphinx-rtd-theme==0.4.3',  # used to build html docs locally
-        'tensorboard == 2.1.0',
+        'tensorboard == 2.6.0',
```
Comment (hnyu) [inline setup.py:47]: Upgrade tensorboard to 2.6 for fast data loading.

File: alf/environments/gym_wrappers.py [hunk 3]
```diff
@@ -689,3 +689,48 @@ def __init__(self, env):
     def step(self, action):
         ob, reward, done, info = self.env.step(action)
         return ob, reward, False, info
+
+
+@alf.configurable
+class DiscreteActionWrapper(gym.ActionWrapper):
```
Comment (emailweixu) [inline alf/environments/gym_wrappers.py]: As we now support some native non-gym batched environment, it's better to make it a batch alf wrapper so it can also be used for those non-gym environments.

File: alf/algorithms/sac_algorithm.py [hunk 4]
```diff
@@ -558,18 +556,59 @@ def rollout_step(self, inputs: TimeStep, state: SacState):
...
+                         apply_reward_weights=True):
         if self._act_type == ActionType.Continuous:
             observation = (observation, action)
         elif self._act_type == ActionType.Mixed:
             observation = (observation, action[1])  # continuous action
         # discrete/mixed: critics shape [B, replicas, num_actions]
         # continuous: critics shape [B, replicas]
         critics, critics_state = critic_net(observation, state=critics_state)
+
+        # For mult-dim reward, do
```
Comment (Haichao-Zhang) [inline alf/algorithms/sac_algorithm.py]: mult-dim -> multi-dim

File: alf/environments/gym_wrappers.py [hunk 5]
```diff
@@ -689,3 +689,48 @@ def __init__(self, env):
...
+    """Discretize each continuous action dim into several evenly distributed
+    values. Currently only support unnested action space as ``gym.spaces.Box``
+    with a rank-1 shape.
+    """
+
+    def __init__(self, env: gym.Env, actions_num: int):
```
Comment (Haichao-Zhang) [inline alf/environments/gym_wrappers.py]: actions_num: the name could be more precise (e.g. bin_num), to reduce ambiguity with action dimensions.

Comment (hnyu) [reply alf/environments/gym_wrappers.py]: > actions_num: the name could be more precise (e.g. bin_num), to reduce ambiguity with action dimensions.

I think the the docstring is clear enough. It's not bin number as this has nothing to do with histogram.

File: alf/environments/gym_wrappers.py (hunk 3 above)
Comment (hnyu) [reply alf/environments/gym_wrappers.py]: Changed to an alf wrapper

**Review by Haichao-Zhang — APPROVED** [review]
